source/usage-examples/findOne.go

```diff
@@ -0,0 +1,47 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+// Replace the uri string with your MongoDB deployment's connection string.
+const uri = "mongodb+srv://<user>:<password>@<cluster-url>?retryWrites=true&w=majority"
+
+func main() {
+	ctx := context.TODO()
+	client, err := mongo.Connect(ctx, options.Client().ApplyURI(uri))
+	if err != nil {
+		panic(err)
+	}
+	defer func() {
+		if err = client.Disconnect(ctx); err != nil {
+			panic(err)
+		}
+	}()
+
+	coll := client.Database("sample_mflix").Collection("movies")
+	var result bson.D
+	opts := options.FindOne().
+		SetProjection(bson.D{{"_id", 0}, {"title", 1}, {"imdb", 1}}).
+		SetSort(bson.D{{"imdb.rating", -1}})
+	err = coll.FindOne(ctx, bson.D{{"title", "The Room"}}, opts).Decode(&result)
+	if err != nil {
+		if err == mongo.ErrNoDocuments {
+			// This error means your query did not match any documents.
+			return
+		}
+		panic(err)
+	}
+
+	jsonByteSlice, err := json.MarshalIndent(result, "", "    ")
```

Comment:
This cannot use `encoding/json` to marshal a `bson.D`. It marshals the `bson.D` type to an array of key/value pair documents:

```

    {
        "Key": "title",
        "Value": "The Room"
    },
    {
        "Key": "imdb",
        "Value": [
            {
                "Key": "rating",
                "Value": 3.5
            },
            {
                "Key": "votes",
                "Value": 25673
            },
            {
                "Key": "id",
                "Value": 368226
            }
        ]
    }
]
```

Instead, I think this should use `bson.MarshalExtJSON`. Note, that it does not indent – that is an open feature request in [GODRIVER-946](https://jira.mongodb.org/browse/GODRIVER-946). But it should produce the output:
```
{"title":"The Room","imdb":{"rating":3.5,"votes":25673,"id":368226}}
```

Reply:
Thanks for the note.  Fix:

To confirm, [`bson.MarshalExtJson`]() 

To help test my understanding of the situation and so that I don't make the same mistake twice, is the following correct:

[`bson.D`](https://pkg.go.dev/github.com/mongodb/mongo-go-driver@v1.2.1/bson#D) is, under the hood, an array of [`primitive.E`](https://pkg.go.dev/go.mongodb.org/mongo-driver/bson/primitive#E) type values (which are structs with a  string `key` field and an `interface{}` `value` field). When `[encoding/json.MarshalIndent](https://golang.org/pkg/encoding/json/#MarshalIndent)` is called on `bson.D`,  it recursively traverses the `bson.D` type value and thus prints out the list of structs above?

Reply:
> canonical is set to false (which as I understand it outputs relaxed mode json)

Correct

> escapeHTML can be set to either true or false as this example does not contain characters that will require HTML escaping as per encoding/json.HTMLEscape() .

Yes, comparing the bson packages [writeStringWithEscapes](https://github.com/mongodb/mongo-go-driver/blob/08107589b91d4ce92fed1c46428b9135861b3a86/bson/bsonrw/extjson_writer.go#L637:L637) it appears to escape the same set of characters that `encoding/json.HTMLEscape` does: `<`, `>`, `&` and the separators U+2028 U+2029.

> To test my understanding and so that I don't make the same mistake twice, is the following explanation of why the encoding/json.MarshalIndent() method is incorrect in this situation accurate? This isn't going in the docs, this is only me personally trying to make sure I have an accurate mental model.

Precisely!